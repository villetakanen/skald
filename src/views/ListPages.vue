<template>
  <v-container>

    <v-row>
      <v-col>
        <TabTitle
          topic="Page list"/>
      </v-col>
    </v-row>

    <v-row>
      <v-col>
        <v-card>
          <v-card-text>
            <v-row>
              <v-col
                cols='4'>
                <h1>All</h1>
                <ul>
                  <template v-for="(page, index) in pageIndex">
                    <li v-bind:key="index">
                      <router-link :to="`/v/${page.path}`">{{page.name}}</router-link></li>
                  </template>
                </ul>
              </v-col>
              <v-col
                cols='4'>
                <h1 v-if="characterIndex.length > 0">Characters</h1>
                <ul>
                  <template v-for="(page, index) in characterIndex">
                    <li v-bind:key="index">
                      <router-link :to="`/v/${page.path}`">{{page.name}}</router-link></li>
                  </template>
                </ul>
              </v-col>
              <v-col
                cols='4'>
                <h1 v-if="characterIndex.length > 0">Game Logs</h1>
                <ul>
                  <template v-for="(page, index) in gameLogIndex">
                    <li v-bind:key="index">
                      <router-link :to="`/v/${page.path}`">{{page.name}}</router-link></li>
                  </template>
                </ul>
              </v-col>
            </v-row>
          </v-card-text>
        </v-card>
      </v-col>
    </v-row>
  </v-container>
</template>

<script>
import TabTitle from '../components/TabTitle'

export default {
  components: {
    TabTitle
  },
  props: [
    'siteid'
  ],
  created () {
    this.updateSite(this.siteid)
  },
  watch: {
    '$route' (to, from) {
      this.updateSite(this.siteid)
    }
  },
  computed: {
    pageIndex () {
      return this.$store.getters['site/pageIndex']()
    },
    characterIndex () {
      // @todo, these should be dynamic at some point maybe :D
      return this.$store.getters['site/pageIndex']('Character')
    },
    gameLogIndex () {
      // @todo, these should be dynamic at some point maybe :D
      return this.$store.getters['site/pageIndex']('Game Log')
    }
  },
  methods: {
    updateSite (siteid) {
      // console.log('siteid', siteid, this.siteid)
      if (siteid === null || typeof siteid === 'undefined') siteid = 'skald'
      this.$store.dispatch('binder/openPage', { siteid: siteid, pageid: siteid })

      // set site styles
      this.$store.dispatch('sites/openSite', siteid)
      this.$store.dispatch('site/open', siteid)
    }
  }
}
</script>
