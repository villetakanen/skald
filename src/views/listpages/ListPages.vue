<template>
  <v-container >
    <v-row v-if="meta.loading">
      <v-col>
        <Loading
          center
          class="ma-4"/>
      </v-col>
    </v-row>
    <v-row>
      <v-col>
        <TabTitle
          :sub="site.name"
          :sublink="`/v/${site.siteid}`"
          :topic="$t('pagelist.title')"/>
        </v-col>
      </v-row>
      <v-row v-if="!meta.loading">
        <v-col>
          <v-card>
            <v-card-text>
              <div class="index">
              <template v-for="(cat, index) in categories">
                <div v-bind:key="index" class="catIndex" v-if="pagesForCategory(cat.value).length > 0">
                  <h1>{{cat.text}}</h1>
                  <template v-for="(page, indexPages) in pagesForCategory(cat.value)">
                    <p v-bind:key="indexPages"><a :href="`/#/v/${page.siteid}/${page.pageid}`">{{page.name}}</a></p>
                  </template>
                </div>
              </template>
              </div>
            </v-card-text>
          </v-card>
        </v-col>
      </v-row>
  </v-container>
</template>

<script lang="ts">
import Vue from 'vue'
import VueCompositionAPI, { defineComponent, onMounted, ref, watch, onUnmounted } from '@vue/composition-api'
import firebase, { firestore } from 'firebase/app'
import 'firebase/firestore'
import Loading from '@/components/Loading.vue'
import TabTitle from '@/components/app/TabTitle.vue'
import { useSite } from '@/lib/useSite'
import { Page } from '@/plugins/skaldfire'
import { usePage } from '@/lib/usePage'
Vue.use(VueCompositionAPI)

export default defineComponent({
  components: {
    Loading,
    TabTitle
  },
  props: {
    siteid: {
      type: [String],
      required: true
    }
  },
  setup (props) {
    const { site, meta } = useSite()
    const { categories } = usePage()
    const listArray:Page[] = []
    const pageList = ref(listArray)
    let unsubscribe = () => {}

    onMounted(() => { subscribe(props.siteid) })
    onUnmounted(() => { unsubscribe() })
    watch(site, (newSite) => {
      console.log(newSite)
      subscribe(newSite.siteid)
    })

    function subscribe (siteid:string):void {
      console.log('opening up', siteid)
      if (!siteid) return
      unsubscribe()
      pageList.value.length = 0

      const db = firebase.firestore()
      const listRef = db.collection('sites').doc(siteid).collection('pages')

      unsubscribe = listRef.onSnapshot((snapshot) => {
        snapshot.docChanges().forEach((docChange) => {
          const data = docChange.doc.data()
          pageList.value.push({
            pageid: docChange.doc.id,
            siteid: siteid,
            name: data.name,
            category: data.category
          })
        })
      })
    }

    function pagesForCategory (cat:string):Page[] {
      if (!cat) {
        return pageList.value.filter((item) => { return !item.category })
      }
      return pageList.value.filter((item) => { return item.category === cat })
    }

    return { site, meta, categories, pageList, pagesForCategory }
  }
})
</script>

<style lang="scss" scoped>
.index {
  display: flex;
  flex-wrap: wrap;
  .catIndex{
    min-width: 220px;
    flex-grow: 1;
    h1 {
      margin-bottom: 16px;
    }
    p {
      margin: 0;
      margin-bottom: 8px;
    }
  }
  .catIndex+.catIndex {
    border-left: solid 1px rgba(0,0,0,0.4);
    padding-left: 8px;
  }
}
</style>
