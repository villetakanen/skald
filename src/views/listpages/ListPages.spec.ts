import { shallowMount } from '@vue/test-utils'
import ListPages from './ListPages.vue'

describe('Mounted App', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(ListPages, {
      mocks: {
        $t: () => { return '' }
      },
      propsData: {
        siteid: 'mekanismi'
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
