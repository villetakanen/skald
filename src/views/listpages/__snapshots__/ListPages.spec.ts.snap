// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mounted App renders correctly 1`] = `
<v-container-stub
  tag="div"
>
  <!---->
   
  <v-row-stub
    tag="div"
  >
    <v-col-stub
      tag="div"
    >
      <tabtitle-stub
        sub=""
        sublink="/v/"
        topic=""
      />
    </v-col-stub>
  </v-row-stub>
   
  <v-row-stub
    tag="div"
  >
    <v-col-stub
      tag="div"
    >
      <v-card-stub
        loaderheight="4"
        tag="div"
      >
        <v-card-text-stub>
          <div
            class="index"
          >
            <!---->
            <!---->
            <!---->
            <!---->
            <!---->
          </div>
        </v-card-text-stub>
      </v-card-stub>
    </v-col-stub>
  </v-row-stub>
</v-container-stub>
`;
